## Clustering Neighbourhoods in Toronto
Coursera Capstone Project - Part 1 & Part 2

Guangli Ma

Part 1 Converting Wikipedia html table into a DataFrame

In [1]:
from bs4 import BeautifulSoup as bsoup
from urllib.request import urlopen as uReq
import requests
import lxml
import pandas as pd
from pandas import DataFrame
import numpy as np

In [2]:
my_url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

In [3]:
r=requests.get(my_url)

## Parsing the web html file with BeautifulSoup package

In [4]:
# Parse the htlm with Soup
page=bsoup(r.text,"html.parser")
page


<!DOCTYPE html>

<html class="client-nojs" dir="ltr" lang="en">
<head>
<meta charset="utf-8"/>
<title>List of postal codes of Canada: M - Wikipedia</title>
<script>document.documentElement.className="client-js";RLCONF={"wgBreakFrames":!1,"wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June","July","August","September","October","November","December"],"wgRequestId":"728a696b-a2dd-4ba4-87fe-2b5f89c0b703","wgCSPNonce":!1,"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":969510799,"wgRevisionId":969510799,"wgArticleId":539066,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Articles with short description","Communications in Ontario","Postal codes in Canada","Toronto","Ontario-rel

## The htlm 'page' indicates that there is a 'table class' - let's find that table

In [5]:
rtable=page.table
rtable

<table class="wikitable sortable">
<tbody><tr>
<th>Postal Code
</th>
<th>Borough
</th>
<th>Neighbourhood
</th></tr>
<tr>
<td>M1A
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M2A
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3A
</td>
<td>North York
</td>
<td>Parkwoods
</td></tr>
<tr>
<td>M4A
</td>
<td>North York
</td>
<td>Victoria Village
</td></tr>
<tr>
<td>M5A
</td>
<td>Downtown Toronto
</td>
<td>Regent Park, Harbourfront
</td></tr>
<tr>
<td>M6A
</td>
<td>North York
</td>
<td>Lawrence Manor, Lawrence Heights
</td></tr>
<tr>
<td>M7A
</td>
<td>Downtown Toronto
</td>
<td>Queen's Park, Ontario Provincial Government
</td></tr>
<tr>
<td>M8A
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M9A
</td>
<td>Etobicoke
</td>
<td>Islington Avenue, Humber Valley Village
</td></tr>
<tr>
<td>M1B
</td>
<td>Scarborough
</td>
<td>Malvern, Rouge
</td></tr>
<tr>
<td>M2B
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3B
</td>
<td

In [6]:
results=rtable.find_all('tr')
nrows=len(results)
nrows

181

In [7]:
results[0:5]

[<tr>
 <th>Postal Code
 </th>
 <th>Borough
 </th>
 <th>Neighbourhood
 </th></tr>, <tr>
 <td>M1A
 </td>
 <td>Not assigned
 </td>
 <td>Not assigned
 </td></tr>, <tr>
 <td>M2A
 </td>
 <td>Not assigned
 </td>
 <td>Not assigned
 </td></tr>, <tr>
 <td>M3A
 </td>
 <td>North York
 </td>
 <td>Parkwoods
 </td></tr>, <tr>
 <td>M4A
 </td>
 <td>North York
 </td>
 <td>Victoria Village
 </td></tr>]

## The Header of the DataFrame is the first row of data [index 0]

In [8]:
header=results[0].text.split()
header

['Postal', 'Code', 'Borough', 'Neighbourhood']

## Let's check some rows in order to prepare to build the loop that will extract all cells into a DataFrame. For example, let's examine row 85

In [9]:
results[85].text

'\nM4M\n\nEast Toronto\n\nStudio District\n'

In [10]:
results[85].text.split('\n')

['', 'M4M', '', 'East Toronto', '', 'Studio District', '']

In [11]:
results[85].text.split('\n')

['', 'M4M', '', 'East Toronto', '', 'Studio District', '']

In [12]:
Postcode=results[85].text.split('\n')[1]
Postcode

'M4M'

In [15]:
Borough=results[85].text.split('\n')[3]
Borough

'East Toronto'

In [16]:
Neighborhood=results[85].text.split('\n')[5]
Neighborhood

'Studio District'

## Iteration loop to extract all cells into a dataframe df

In [18]:
# iteration loop to harvest all records

records =[]
n=1
while n < nrows :
    Postcode=results[n].text.split('\n')[1]
    Borough=results[n].text.split('\n')[3]
    Neighborhood=results[n].text.split('\n')[5]
    records.append((Postcode, Borough,Neighborhood))
    n=n+1

df=pd.DataFrame(records, columns=['PostalCode', 'Borough', 'Neighbourhood'])
df.head(5)

PostalCode           Borough              Neighbourhood
0        M1A      Not assigned               Not assigned
1        M2A      Not assigned               Not assigned
2        M3A        North York                  Parkwoods
3        M4A        North York           Victoria Village
4        M5A  Downtown Toronto  Regent Park, Harbourfront

## Initial Dataframe has 289 cells or rows of data, as expected.

In [19]:
df.shape

(180, 3)

## Let's check the tail of dataframe to assure all data is included as in Wikipedia table

In [20]:
df.tail()    #  Perfect! all data is included.

PostalCode       Borough  \
175        M5Z  Not assigned   
176        M6Z  Not assigned   
177        M7Z  Not assigned   
178        M8Z     Etobicoke   
179        M9Z  Not assigned   

                                         Neighbourhood  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned

## Now, let's remove all rows with Borough = 'Not assigned' strings
There are 77 Boroughs with 'Not assigned' strings

In [21]:
# How many rows have Borough equal to 'Not assigned'?
df[df['Borough']=='Not assigned'].count()

PostalCode       77
Borough          77
Neighbourhood    77
dtype: int64

In [22]:
# drops those rows where 'Not assigned' appears in column '[Borough]'
df1=df[~df.Borough.str.contains("Not assigned")]
df1=df1.reset_index(drop=True)

In [23]:
df1.head(5)

PostalCode           Borough                                Neighbourhood
0        M3A        North York                                    Parkwoods
1        M4A        North York                             Victoria Village
2        M5A  Downtown Toronto                    Regent Park, Harbourfront
3        M6A        North York             Lawrence Manor, Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [24]:
df1.tail()

PostalCode           Borough  \
98         M8X         Etobicoke   
99         M4Y  Downtown Toronto   
100        M7Y      East Toronto   
101        M8Y         Etobicoke   
102        M8Z         Etobicoke   

                                         Neighbourhood  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...

## The dataframe now has 180 rows (0-179) after droppong all Borough with 'Not assigned', which is correct ( 180-77)

In [25]:
df1.shape

(103, 3)

## Now, let's replace 'Not assigned' neighborhoods with the name of the Borough

In [26]:
df1.loc[df1['Neighbourhood'] == 'Not assigned', 'Neighbourhood'] = df1['Borough']

In [27]:
df1.head()           # Now, there are no rows with 'Not assigned' strings anymore

PostalCode           Borough                                Neighbourhood
0        M3A        North York                                    Parkwoods
1        M4A        North York                             Victoria Village
2        M5A  Downtown Toronto                    Regent Park, Harbourfront
3        M6A        North York             Lawrence Manor, Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [28]:
df1.shape

(103, 3)

## Now , we will consolidate the dataframe to the unique Postal Codes.
As per instructions, we will look for repeated postalcodes and consolidate into one postalcode with the name of the neighbourhoods merged.
Let's examine de dataframe first to set up the best loop algorithm. How many unique postalcodes , Boroughs and Neighbourhood names are there?

In [29]:
postalcodes = df1['PostalCode'].nunique()
boroughs = df1['Borough'].nunique()
neighbourhoods= df1['Neighbourhood'].nunique()
print('Unique Postalcodes : ' + str(postalcodes))
print('Unique Boroughs  : '+ str(boroughs))
print('Unique Neighbourhoods  :' + str(neighbourhoods))

Unique Postalcodes : 103
Unique Boroughs  : 10
Unique Neighbourhoods  :99


## There are 103 Unique Postcodes and 11 Unique Borough with 210 Neighbourhoods

In [30]:
df1.groupby(['PostalCode','Borough','Neighbourhood']).size().reset_index(name='Count').head()

PostalCode      Borough                           Neighbourhood  Count
0        M1B  Scarborough                          Malvern, Rouge      1
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek      1
2        M1E  Scarborough       Guildwood, Morningside, West Hill      1
3        M1G  Scarborough                                  Woburn      1
4        M1H  Scarborough                               Cedarbrae      1

## Let's Consolidate the dataframe to each unique PostalCodes and aggregated Neighbourhoods

In [31]:
# Starting DataFrame df1
df1.head()

PostalCode           Borough                                Neighbourhood
0        M3A        North York                                    Parkwoods
1        M4A        North York                             Victoria Village
2        M5A  Downtown Toronto                    Regent Park, Harbourfront
3        M6A        North York             Lawrence Manor, Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [32]:
nrows1=len(df1)
nrows1

103

## Let's keep df1 intact and let's work with df2 as precaution

In [33]:
# DataFrame df2 also has 212 rows or cells
df2=df1
df2.head()

PostalCode           Borough                                Neighbourhood
0        M3A        North York                                    Parkwoods
1        M4A        North York                             Victoria Village
2        M5A  Downtown Toronto                    Regent Park, Harbourfront
3        M6A        North York             Lawrence Manor, Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [34]:
df2.index

RangeIndex(start=0, stop=103, step=1)

In [35]:
# I will use two pointers, 'n and m' to sweep through the dataframe. So 'n' will vary from
#  0 to 211, and 'm ' will vary from 1 to 212
nrows2=len(df2)-1
nrows2

102

## Loop Algorithm to extract the consolidate dataframe df2 with unique postalcodes.
We expect a df2 with 103 rows because that is the number of unique postalcodes

In [36]:
n=0

while n < nrows2 :
    post1=df2.iloc[n,0]
    #post1
    m=n+1
    post2=df2.iloc[m,0]
    #post2
    neigh1=df2.iloc[n,2]
    neigh2=df2.iloc[m,2]
    if post1==post2:
        df2.Neighbourhood[n,2] = neigh1=neigh1+','+neigh2
        #df2 = df2[df2.Neighbourhood != 'neigh2']
        df2=df2.drop(df2.index[m])
        nrows2=nrows2-1
        df2 = df2.reset_index(drop=True)
    else:
        n=n+1


#df2 = df2.reset_index(drop=True)
# print(post1, post2, 'Nigh1 is...'+ neigh1,',,,,,,,','Nigh2 is...' + neigh2, n,m)
df2.index

RangeIndex(start=0, stop=103, step=1)

In [37]:
df2.head()

PostalCode           Borough                                Neighbourhood
0        M3A        North York                                    Parkwoods
1        M4A        North York                             Victoria Village
2        M5A  Downtown Toronto                    Regent Park, Harbourfront
3        M6A        North York             Lawrence Manor, Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

## DataFrame consolidated to unique 103 PostalCodes

In [38]:
df2.head(15)

PostalCode           Borough  \
0         M3A        North York   
1         M4A        North York   
2         M5A  Downtown Toronto   
3         M6A        North York   
4         M7A  Downtown Toronto   
5         M9A         Etobicoke   
6         M1B       Scarborough   
7         M3B        North York   
8         M4B         East York   
9         M5B  Downtown Toronto   
10        M6B        North York   
11        M9B         Etobicoke   
12        M1C       Scarborough   
13        M3C        North York   
14        M4C         East York   

                                        Neighbourhood  
0                                           Parkwoods  
1                                    Victoria Village  
2                           Regent Park, Harbourfront  
3                    Lawrence Manor, Lawrence Heights  
4         Queen's Park, Ontario Provincial Government  
5             Islington Avenue, Humber Valley Village  
6                                      Malvern, Rouge  
7                                           Don Mills  
8                     Parkview Hill, Woodbine Gardens  
9                            Garden District, Ryerson  
10                                          Glencairn  
11  West Deane Park, Princess Gardens, Martin Grov...  
12             Rouge Hill, Port Union, Highland Creek  
13                                          Don Mills  
14                                   Woodbine Heights

## Dataframe has 103 rows and 3 columns

In [40]:
df2.shape

(103, 3)